### PR TITLE
Font Library: Improve usability of font variant selection

### PR DESCRIPTION
**Pull request by mikachan**
## What?

This updates the font variant markup so that the checkbox controls are wrapped inside a label element. This means that any part of the font variant panel can be clicked to interact with the checkbox, rather than only the checkbox itself.

## Why?

This improves the usability of these checkboxes, as you can click on a much larger area for the toggle interactions. It also should improve their accessibility, as previously these checkboxes did not include any labels. Now when a checkbox is read out by a screenreader, it will read out the font variant name.

## How?

There are two main updates: updating the markup of the font variant selection for both the library and collection versions; and updating the `CheckboxControl` component to include the option for preventing a label from being rendered, in order to allow for a custom label to be used outside the component.

## Testing Instructions

1. Ensure you're using a theme that includes at least one bundled font family (or upload a font family as part of this test)
2. Open the Font Library modal
3. Open a font family
4. Toggle a font variant of the selected font family by clicking anywhere in the font variant panel, rather than just the checkbox

### Testing Instructions for Keyboard

1. Ensure you're using a theme that includes at least one bundled font family (or upload a font family as part of this test)
2. Open the Font Library modal
3. Open a font family
4. Toggle a font variant of the selected font family by tabbing to the font variant panel and toggling the checkbox. If using a screenreader, the font variant name should be called out.

## Screenshots or screencast 

| Before | After |
| ------ | ------ |
| <video src="https://github.com/WordPress/gutenberg/assets/1645628/089895ec-f222-4883-8477-534fe560678d" /> | <video src="https://github.com/WordPress/gutenberg/assets/1645628/da7abdf4-20ae-449c-8e7a-321adc348fa2" /> |

### CODE REVIEW

File: packages/components/src/checkbox-control/index.tsx [hunk 1]
```diff
@@ -125,12 +125,14 @@ export function CheckboxControl(
 					/>
 				) : null }
 			</span>
-			<label
-				className="components-checkbox-control__label"
-				htmlFor={ id }
-			>
-				{ label }
-			</label>
+			{ label !== false && (
```
Comment (jffng) [inline packages/components/src/checkbox-control/index.tsx]:
```suggestion
			{ label && (
```

Comment (jffng) [reply packages/components/src/checkbox-control/index.tsx]: Maybe it's cleaner this way?

**Review by jffng — APPROVED** [review]
Hey! This is testing well for me. I tried both the mouse and keyboard behavior, verified it's working as expected as well as no label is rendered in the Checkbox Control component: 

https://github.com/WordPress/gutenberg/assets/5375500/c318ab12-44df-4dc2-8daa-69b0fc7d9c8c

I left one small suggestion, other than that LGTM.

**Comment by jffng** [comment]
Oh one more thing I thought of when reviewing this PR — should we open an issue for more UX improvements? In particular, it could be useful to have a Select / Unselect all variants option.

File: packages/components/src/checkbox-control/index.tsx (hunk 1 above)
Comment (mikachan) [reply packages/components/src/checkbox-control/index.tsx]: Ah yes it is, thanks!

**Comment by mikachan** [comment]
> should we open an issue for more UX improvements? In particular, it could be useful to have a Select / Unselect all variants option.

Yep, good idea! Opened an issue here: https://github.com/WordPress/gutenberg/issues/56688

**Comment by afercia** [comment]
Noting that the WordPress Accessibility standards recommend to not use wrapping labels, see https://make.wordpress.org/accessibility/handbook/markup/core-accessibility-standards/#labeling

This also causes a double labeling, as the label element still uses a `for` attribute pointing to the input `id`. The editor components come with built-in labeling mechanisms that should not be altered this way. For the future, please avoid altering the default components behavior. We tried to build them with built-in accessibility and to conform to the WP Accessibility best practices for good reasons.

Reported at https://github.com/WordPress/gutenberg/issues/58299

**Comment by afercia** [comment]
While I do understand the good intent behind this change, I'm afraid I _have_ to remark that:
- The guidelines of this project ask to [create an issue for any non-trivial pull request](https://github.com/WordPress/gutenberg/blob/trunk/docs/contributors/repository-management.md#pull-requests), so that the issue _defines the problem to solve and allows for discussion of the most appropriate solution before actually writing code_. However, I couldn't find an issue associated with this pull request.
- This pull request touches an important aspect of accessibility but it's not labeled with any of the two relevant GitHub labels: `[Focus] Accessibility` or `Needs Accessibility Feedback`.

For the greater good of the WordPress project I'm not sure pull requests that do not follow this project guidelines should be allowed.

**Review by ciampo — COMMENTED** [review]
Hey there 👋 

Echoing @afercia 's message, we should be careful when making changes to public components APIs, especially when those changes have the potential to affect the component's accessibility negatively (for example, the snapshot diff in this PR already showed that the `input` in the test ended up unlabeled)

While the original intention behind these changes is understandable, it is now possible for consumers of `CheckboxControl` to avoid setting a `label` altogether, which goes against the [efforts](https://github.com/WordPress/gutenberg/issues/51740) we've been making to improve controls accessibility.

Were any alternative approaches discussed?

File: packages/components/src/checkbox-control/types.ts [hunk 2]
```diff
@@ -19,9 +19,10 @@ export type CheckboxControlProps = Pick<
 	/**
 	 * A label for the input field, that appears at the side of the checkbox.
 	 * The prop will be rendered as content a label element. If no prop is
-	 * passed an empty label is rendered.
+	 * passed an empty label is rendered. If the prop is set to false no label
+	 * is rendered.
 	 */
-	label?: string;
+	label?: string | false;
```
Comment (ciampo) [inline packages/components/src/checkbox-control/types.ts:25]: We usually try to avoid mixing different primitive types on the same prop, where possible.

**Comment by mikachan** [comment]
Hi @afercia and @ciampo, thanks for the feedback here. The intention here was only to improve the accessibility of the checkboxes in the font library, but I understand the points you've raised. It sounds like we need to revert the changes to the component, but we'll likely need to redesign the font library checkboxes in order for them to be accessible again. I'll add this work to the [font library tracking issue](https://github.com/WordPress/gutenberg/issues/55277).

**Comment by afercia** [comment]
> we'll likely need to redesign the font library checkboxes in order for them to be accessible again.

I'm not sure big changes are really needed. We just need to make sure the label is correctly associated to the checkbox. Please do not alter the component default behaevior.

A correctly associated label is natively clickable. It's built-in accessibility.
Making the entire row clickable _may_ be considered an _usability_ improvement but it comes with its own downsides, for example the inability to select text with the mouse because that would trigger a click. When trying to alter native behaviors there's often (always?) some side effect. I wouldn't recommend to make the whole row clickable.

**Pull request merged**